SCAL-216702 added readme file for example and made modification in main READ.md (#53)
thoughtspot/ts-chart-sdk: example/custom-bar-chart/README.md

## Integrate Bar Charts through ts-chart-sdk 

In this tutorial, we will walk you through the process of creating custom bar charts using the `Charts.js` library and integrating them with the TS-Chart-SDK. By the end of this tutorial, you will be able to create interactive bar charts that leverage ThoughtSpot's powerful capabilities, such as data drilling and embedding data into your application. Let's get started!

To integrate ThoughtSpot Chart SDK and complete these steps:

1. [Setup your environment](#setup-your-environment)
2. [Implementing sample bar chart](#implementing-sample-bar-chart)
3. [Intailizing chart context with ts-chart-sdk](#intialize-chart-context)
4. [Create DataModel for input data](#create-a-data-model-for-input-data)
5. [Implemeting the renderChart](#implementing-the-renderChart)
6. [Integrate to ThoughtSpot](#integrate)

### Setup Your Environment

- For this u can refer to [setup your environment](../../README.md#set-up-your-environment) above section with the following changes:
    - While creating folder name it with **bar-chart**.

>     - Give project name as **custom-bar-charts** or u can directly us `npm create vite@latest custom-bar-chart -- --template vanilla-ts`

Reviewer: change 'u' to 'You' everywhere

    - Instead of highcharts install your chart library (which in this case is `charts.js`) using command-`npm i chart.js` to install chart.js and `npm install chartjs-plugin-datalabels` to install datalabels plugin.
> NOTE: 
>For more information about the chart.js and chartjs-plugin-datalabels, refer to the following documentation resources respectively:
  >  * [chart.js getting started](https://www.chartjs.org/docs/latest/getting-started/)

>   > * [chart-plugin-datalabels](https://chartjs-plugin-datalabels.netlify.app/guide/#table-of-contents)

Reviewer: Add a note that the naming of folders does not affect functionality it is just advised to use this nomenclaturem, you can name file or folder according to Your choice,

Author: ack

### Implementing Sample Bar Chart

> In this section we will be Rendering a sample leaflet chart in the application created from the preceding steps.

Reviewer: ```suggestion
In this section we will be Rendering a sample bar chart in the application created from the preceding steps.
```

To implement the chart code in your application, complete these steps:

1. Delete the unwanted files from your project folder.These are metioned as follow-
```
/public [whole folder]
/src/counter.ts
/src/typescripts.svg
/src/style.css
```

> 2. Clear `main.ts` and rename it to `custom-charts.ts`

Reviewer: ```suggestion
2. Clear `main.ts` and rename it to `custom-charts.ts`. This step is not necessary but we advice this nomenclature of files.
```

* [`getQueriesFromChartConfig (Doc)`](#getQueriesFromChartConfig-doc)
* [`validateConfig (Doc)`](#validateConfig-doc)
* [`chartConfigEditorDefinition (Doc)`](#chartConfigEditorDefinition-doc)
* [`visualPropEditorDefinition (Doc)`](#visualPropEditorDefinition-doc)
* [`renderChart (Doc)`](#renderChart-doc)

#### getDefaultChartConfig (Doc)

This function takes in a [ChartModel](https://ts-chart-sdk-docs.vercel.app/interfaces/ChartModel.html) object and returns a well-formed point configuration definition.

Here we are creating geo map for US county data so we will get to column with one column as attribute and other as measure so we are classifying them as `x` and `y` labels.

To create a Leaflet version of the data set, the above-mentioned headers must be presented as columns from ThoughtSpot. The query on the ThoughtSpot Answer page should have all the above columns to plot a Gantt chart.

#### getQueriesFromChartConfig (Doc)

This method defines the data query that is required to fetch the data from ThoughtSpot to render the chart. For most use cases, you do not require the data outside of the columns listed in your chart.

> This example maps all the columns in the configuration as an array of columns in the arguments.

Reviewer: I think some of the above points are only valid for leaflet which you have added in custom bar chart please rectify this.

#### validateConfig (Doc)

> This method use to apply any custom validation that is reqired by developer on the chartconfig and chartModel. For example you want to user have a certain input validation in some form that changing chartconfig or chartModel you can validations here. This a optional method.

Reviewer: required

Author: ack

In the code snippet we are checking if config is not empty.You can send your custom error message in `validationErrorMessage`.

#### chartConfigEditorDefinition (Doc)
This is a optional method provided in getChartContext where you can define the cutom config editor that will be shown in settings in TS and leverage TS feature from there.It take `chartModel` and `customChartContext` as parameter and return the `chartConfigEditorDefinition[]` which contain `columnSection` where you can deifne editor column for each config axis.

> 

Reviewer: always add space after full stop

> This is a optional method provided in getChartContext where you can define the cutom config editor that will be shown in settings in TS and leverage TS feature from there. It take `chartModel` and `customChartContext` as parameter and return the `chartConfigEditorDefinition[]` which contain `columnSection` where you can deifne editor column for each config axis.

Reviewer: ```suggestion
This is a optional method provided in getChartContext where you can define the custom config editor that will be shown in settings in TS and leverage TS feature from there. It take `chartModel` and `customChartContext` as parameter and return the `chartConfigEditorDefinition[]` which contain `columnSection` where you can define editor column for each config axis.
```

> NOTE: 
>For more information about the ChartToTSEvents component, refer to the following documentation resources:
  >  * [https://ts-chart-sdk-docs.vercel.app/interfaces/ChartToTSEventsPayloadMap.html](https://ts-chart-sdk-docs.vercel.app/interfaces/ChartToTSEventsPayloadMap.html)
```ts
const renderChart = async (ctx: CustomChartContext): Promise<void> => {
  try {
      ctx.emitEvent(ChartToTSEvent.RenderStart);
      render(ctx);
  } catch (e) {
      ctx.emitEvent(ChartToTSEvent.RenderError, {
          hasError: true,
          error: e,
      });
  } finally {
      ctx.emitEvent(ChartToTSEvent.RenderComplete);
  }
};                
```

## Integrate to thoughtspot

In this sectiom we will be working on integrating your custom chart with thoughtspot. For integrating do the following steps->

1. Deploy this custom app on any hosting services currenly i am using ngrok to local take a url with ngrok you will get url `https://<random_string>.ngrok-free.app`
2. Redeploy the cluster with the correct flag settings.
2. Open thoughtspot go `admin` go to `chart customization` then go to `custom charts`   

> 3. Click on `create charts` then name it. We are going with `custom-bar-chart` for current example put the url where you deployed the app for example:``https://<random_string>.ngrok-free.app``.

Reviewer: specify this is for local development and to add charts to cluster this needs to be deployed to a service preferrably vercel

Author: ack
